In Julia 1.3 lock condition before notify() or wait() on it

quatrix opened the pull request.

quatrix commented:
#487

DilumAluthge commented:
Instead of doing
```
lock(cond)
... stuff ...
unlock(cond)
```

Perhaps it would be better to do:
```
lock(cond) do
    ... stuff...
end
```

quatrix commented:
@DilumAluthge Thanks! changed to lock() do

quinnj commented:
@quatrix, can you show some tests or how this solves #487? I know it's tricky to test threaded scenarios on travis, so I'm not necessarily saying we need more tests, but at least something a few of us can try locally before/after these changes to get a feel for what's getting solved.

quatrix commented:
@quinnj, actually you can try the example in #487 it works with this PR and throws an exception with the master when running with ```JULIA_NUM_THREADS=2```

Maybe it's an overkill but I've added that as part of the ```client.jl``` tests, calling GET/POST/PUT/etc inside a ```Threads.@spawn``` and added to ```.travis.yml``` to run the tests with ```JULIA_NUM_THREADS=1``` and ```JULIA_NUM_THREADS=2```

NHDaly commented on src/ConnectionPool.jl:
```diff
@@ -240,7 +262,7 @@ function IOExtras.startwrite(t::Transaction)
-        wait(t.c.writedone)
+        _wait(t.c.writedone)
     end                                           ;@debug 2 "👁  Start write:$t"
     t.c.writebusy = true
```
You almost certainly want to be holding the lock around this entire section, not just within the `wait` function itself. Note this example on wikipedia about how to use a Monitor (a combo Condition Variable + Mutex, which is what Julia's `Condition()` type provides):
https://en.wikipedia.org/wiki/Monitor_(synchronization)#Monitor_usage
```
acquire(m); // Acquire this monitor's lock.
while (!p) { // While the condition/predicate/assertion that we are waiting for is not true...
	wait(m, cv); // Wait on this monitor's lock and condition variable.
}
// ... Critical section of code goes here ...
signal(cv2); -- OR -- notifyAll(cv2); // cv2 might be the same as cv or different.
release(m); // Release this monitor's lock.
```

You'll need to lock during accesses to `writebusy`, `writecount`, `sequence`, etc, here, lest they be updated simultaneously from multiple threads.

It might be that this entire function, actually, should be within the lock.

NHDaly commented on src/ConnectionPool.jl:
```diff
@@ -257,8 +279,8 @@ function IOExtras.closewrite(t::Transaction)
+    _notify(poolcondition[Threads.threadid()])
 
     @ensure !iswritable(t)
     return
```
Same here; it looks to me like this whole section should be `lock`ed, not just within the `_notify` function itself.

NHDaly commented on src/ConnectionPool.jl:
```diff
@@ -274,7 +296,7 @@ function IOExtras.startread(t::Transaction)
     t.c.timestamp = time()
     while t.c.readcount != t.sequence
-        wait(t.c.readdone)
+        _wait(t.c.readdone)
```
and here, etc

quinnj commented:
superseded by on-going work in https://github.com/JuliaWeb/HTTP.jl/pull/509
